stordco/buffy
# feat: SIGNAL-5504: add optional timer based work

SIGNAL-5504

There's a [need](https://stord-team.slack.com/archives/C028DNC1UTW/p1702398162897429?thread_ts=1701980630.485939&cid=C028DNC1UTW) to have Buffy operate in a different flow than original.

Instead of changing the existing flow, a new behavior + using macro was created. Key differences:

  It keeps the following functionality:
    - Wait for a specified amount of time before
  invoking the work function. If the function is called again before the time has
  elapsed, it's a no-op.

Key differences:
    - it will not be terminated once the timer is done, but kept alive
      - internally, the existing timer behavior is done via state rather than handling `{:error, {:already_started, pid}}` output of `GenServer.start_link`.
        -  Under Horde, the state unfortunately doesn't get synced up automatically - that requires explicit tooling.
  Therefore state will be "reset" to the initial state when process boots up. This is not a big issue as the initial state is to
  set a throttled run of `handle_throttle`.
    - it will be given the option (set by config) to trigger work repeatedly based on a empty inbox timeout interval,
      that is based on [GenServer's timeout feature](https://hexdocs.pm/elixir/1.12/GenServer.html#module-timeouts).

### How to Locally Test:

You need to clone this and in `wms-service`'s `mix.exs`, use the local copy for [dependency](https://hexdocs.pm/mix/Mix.Tasks.Deps.html). You might need to also build this library (no need to build this library). 
In `deps`, change the `:buffy` import statement to something like:
```
{:buffy, path: "/Users/[username]/path/to/buffy"},
```
(if you need to know the full path to buffy, do `pwd` in whereever you git cloned this repo)

Also, in `wms-service`, do `rm -rf deps && rm -rf _build && mix deps.get` for safe measure (might not need to do so), in the off chance that released `buffy` is lingering around.

Then, change the `FulfillableStatusCheckBuffer`:
```
defmodule WarehouseEvents.FulfillableStatusCheckBuffer do
  use Buffy.ThrottleAndTimed, #<-- here
    registry_module: Horde.Registry,
    registry_name: Warehouse.HordeRegistry,
    supervisor_module: Horde.DynamicSupervisor,
    supervisor_name: Warehouse.DistributedSupervisor,
    throttle: Application.get_env(:warehouse, :fulfillable_status_check_throttle),
    loop_interval: 2_000 # <-- 2 seconds 

  require Logger
``` 

To test/make the timed interval triggers run *without* an initial order creation (as in when application starts), do something along the lines of:

Although local test seems to show there's some code that triggers `throttle()`(see loom), we want to explicitly start a timed interval in the final implementation. We need to create a child spec for the application Supervisor (typically in `application.ex`) for a Task module, that does a one-off run to start however many instances of `throttle/1` as necessary. That way, the default inbox timeout will run even without any triggers. Following code can be wrapped up in a module, but for testing you can do some variation of:

  ```
  # application.ex
  def start(_type, _args) do
    ...
    children = [
      ...
      # Start timed interval run of WarehouseEvents.FulfillableStatusCheckBuffer for all valid shippers
      {true,
       Supervisor.child_spec(
         {Task,
          fn ->
            building_id = "ac61643b-8eed-4aa4-a377-1e197a91733d"
            building = Warehouse.Repo.get_by(Warehouse.Facility.Building, id: building_id)
            user = Warehouse.Facility.User.system_user(building)

            %{entries: shippers} =
              Warehouse.Facility.list_shippers(%{"buiding_ids" => building_id, "page_size" => 1000}, user)

            for s <- shippers do
              WarehouseEvents.FulfillableStatusCheckBuffer.throttle({building, s.id})
            end
          end},
         id: FulfillableStatusCheckBufferInit,
         restart: :temporary
       )}
    ]
    ...
  ```

### Loom going over the above changes needed for local testing, and showing the current / new behavior differences:

https://www.loom.com/share/e65917c74fe846529af4f2c226abfba2
